### ISSUE TITLE
SPDA doc details, difference between kernels

### ISSUE THREAD

**Issue by lendle** [pytorch/pytorch#99253]
### 📚 The doc issue

The [spda](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) docs do not clearly describe the differences and tradeoffs of the three available SDPA kernels. I had a bunch of questions which I've mostly figured out at this point. Some suggested points to add are included below. 

### Suggest a potential alternative/fix

I think it would be helpful to at least include the following points (assuming they are correct): 

* Describe the memory and compute differences between the three algorithms. I believe the most important is the niave implementation is O(seq_len^2) memory but Flash and Memory Efficient are O(seq_len). According to [the Flash Attention](https://arxiv.org/pdf/2205.14135.pdf) paper, section 5.B, Flash is also much faster, but Memory Efficient is about the same as the naive implementation.
* Link to https://arxiv.org/abs/2112.05682 rather than xformers (or in addition to) for Memory efficient attention
* The precedence of kernels, if all are available for a certain setting. Docs just say it tries to pick the "most optimal". (I think it's Flash -> Memory Efficient -> naive?)
* The docs say "Each of the fused kernels has specific input limitations" but doesn't explicitly explain that Flash and Memory Efficient kernels are considered "fused", and doesn't describe what those limits are.
* It's a bit weird that one of the implementations is called the "Pytorch implementation". Aren't they all, at this point?
* Maybe link directly to [this tutorial](https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html#conclusion) which is a bit more detailed, assuming it's up to date.

cc @svekars @carljparker @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by vadimkantorov** [comment]
related questions: https://github.com/pytorch/pytorch/issues/96968

**Comment by gchanan** [comment]
CC @drisspg

**Comment by jbschlosser** [comment]
Triage review to discuss SDPA label

**Comment by mikekgfb** [comment]
> The precedence of kernels, if all are available for a certain setting. Docs just say it tries to pick the "most optimal". (I think it's Flash -> Memory Efficient -> naive?)

That's what it is today.  There's no commitment that it will always be, so it should not be a part of the documentation.  The documented approach is that we pick one, and you should call the general sdpa operator.

> The docs say "Each of the fused kernels has specific input limitations" but doesn't explicitly explain that Flash and Memory Efficient kernels are considered "fused", and doesn't describe what those limits are.

Purposely so.  You should not call the various implementation functions directly.  There's no commitment that that those will be preserved.  There's only one stable interface, scaled_dot_product_attention.  Users should not rely on specific behaviors of possible implementations for program correctness.  

> It's a bit weird that one of the implementations is called the "Pytorch implementation". Aren't they all, at this point?

In some sense, yes.. You might rephrase it "This implementation provides the  full implementation of all canonical capabilities of the scaled_dot_product_attention operator.  Other implementations provide a subset of those features and are called when calling them appears to be advantageous.  Other implementations are not considered a specified interface of PyTorch, and limitations and specifications may be changed, or the functions may be removed in the future entirely.".

**Comment by vadimkantorov** [comment]
As an argument for more explanation: many of pytorch users are quite advanced and deserve to know the current implementation / dispatch details even if they are subject to change in the future

In practice, sometimes one would optimize the inputs/arch specifically to have the better kernel selected

**Comment by mikekgfb** [comment]
We understand the desire to know more about the implementation which is why we published this blog to add more context and information => https://pytorch.org/blog/accelerated-pytorch-2/

For the most advanced users, the details may be derived from the source code which is available, as well as from the blog we have published.  Documenting the limitations as part of the interface might be misleading, because it might be construed as representing an interface committment.

_scaled_dot_product_flash_attention  and _scaled_dot_product_efficient_attention are internal functions and not documented as part of the Pytorch API.  In effect, they don't exist as part of the committed API.  There are many issues surrounding these custyom kernels - like what SKUs they work on, and to what extent because it's practically impossible for us to test these kernels on all SKUs - we only have access to a few data center SKUs. Thus, these boundaries are actually a function of even the hardware they run on -- you'll find issues around kernels not working with specific inputs on particular consumer hardware SKUs that we could not test on.
